cpp/src/arrow/compute/kernels/aggregate_basic.cc

```diff
@@ -336,8 +336,8 @@ struct ProductImpl : public ScalarAggregator {
       internal::VisitArrayValuesInline<ArrowType>(
           data,
           [&](typename TypeTraits<ArrowType>::CType value) {
-            this->product =
-                MultiplyTraits<AccType>::Multiply(*out_type, this->product, value);
+            this->product = MultiplyTraits<AccType>::Multiply(
+                *out_type, this->product, static_cast<ProductType>(value));
```

Comment:
FTR, why was it necessary to add these casts?

Reply:
Because the constructors for Decimal32/Decimal64 are marked `explicit` I needed to explicitly add casts there to satisfy the templates for conversions